// @warpgatex/warp-smart-order-router/build/main/util/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nativeOnChain = exports.ExtendedEther = exports.WRAPPED_NATIVE_CURRENCY = exports.ID_TO_PROVIDER = exports.CHAIN_IDS_LIST = exports.ID_TO_NETWORK_NAME = exports.NATIVE_CURRENCY = exports.NATIVE_NAMES_BY_ID = exports.NativeCurrencyName = exports.ChainName = exports.ID_TO_CHAIN_ID = exports.NETWORKS_WITH_SAME_UNISWAP_ADDRESSES = exports.HAS_L1_FEE = exports.V2_SUPPORTED = exports.SUPPORTED_CHAINS = void 0;
const warp_sdk_core_1 = require("@warpgatex/warp-sdk-core");
// WIP: Gnosis, Moonbeam
exports.SUPPORTED_CHAINS = [
    warp_sdk_core_1.ChainId.MAINNET,
    warp_sdk_core_1.ChainId.OPTIMISM,
    warp_sdk_core_1.ChainId.OPTIMISM_GOERLI,
    warp_sdk_core_1.ChainId.ARBITRUM_ONE,
    warp_sdk_core_1.ChainId.ARBITRUM_GOERLI,
    warp_sdk_core_1.ChainId.POLYGON,
    warp_sdk_core_1.ChainId.POLYGON_MUMBAI,
    warp_sdk_core_1.ChainId.GOERLI,
    warp_sdk_core_1.ChainId.SEPOLIA,
    warp_sdk_core_1.ChainId.CELO_ALFAJORES,
    warp_sdk_core_1.ChainId.CELO,
    warp_sdk_core_1.ChainId.BNB,
    warp_sdk_core_1.ChainId.AVALANCHE,
    warp_sdk_core_1.ChainId.BASE,
    // Gnosis and Moonbeam don't yet have contracts deployed yet
];
exports.V2_SUPPORTED = [warp_sdk_core_1.ChainId.MAINNET, warp_sdk_core_1.ChainId.GOERLI, warp_sdk_core_1.ChainId.SEPOLIA];
exports.HAS_L1_FEE = [
    warp_sdk_core_1.ChainId.OPTIMISM,
    warp_sdk_core_1.ChainId.OPTIMISM_GOERLI,
    warp_sdk_core_1.ChainId.ARBITRUM_ONE,
    warp_sdk_core_1.ChainId.ARBITRUM_GOERLI,
    warp_sdk_core_1.ChainId.BASE,
    warp_sdk_core_1.ChainId.BASE_GOERLI,
];
exports.NETWORKS_WITH_SAME_UNISWAP_ADDRESSES = [
    warp_sdk_core_1.ChainId.MAINNET,
    warp_sdk_core_1.ChainId.GOERLI,
    warp_sdk_core_1.ChainId.OPTIMISM,
    warp_sdk_core_1.ChainId.ARBITRUM_ONE,
    warp_sdk_core_1.ChainId.POLYGON,
    warp_sdk_core_1.ChainId.POLYGON_MUMBAI,
];
const ID_TO_CHAIN_ID = (id) => {
    switch (id) {
        case 1:
            return warp_sdk_core_1.ChainId.MAINNET;
        case 5:
            return warp_sdk_core_1.ChainId.GOERLI;
        case 11155111:
            return warp_sdk_core_1.ChainId.SEPOLIA;
        case 56:
            return warp_sdk_core_1.ChainId.BNB;
        case 10:
            return warp_sdk_core_1.ChainId.OPTIMISM;
        case 420:
            return warp_sdk_core_1.ChainId.OPTIMISM_GOERLI;
        case 42161:
            return warp_sdk_core_1.ChainId.ARBITRUM_ONE;
        case 421613:
            return warp_sdk_core_1.ChainId.ARBITRUM_GOERLI;
        case 137:
            return warp_sdk_core_1.ChainId.POLYGON;
        case 80001:
            return warp_sdk_core_1.ChainId.POLYGON_MUMBAI;
        case 42220:
            return warp_sdk_core_1.ChainId.CELO;
        case 44787:
            return warp_sdk_core_1.ChainId.CELO_ALFAJORES;
        case 100:
            return warp_sdk_core_1.ChainId.GNOSIS;
        case 1284:
            return warp_sdk_core_1.ChainId.MOONBEAM;
        case 43114:
            return warp_sdk_core_1.ChainId.AVALANCHE;
        case 8453:
            return warp_sdk_core_1.ChainId.BASE;
        case 84531:
            return warp_sdk_core_1.ChainId.BASE_GOERLI;
        case 13473:
            return warp_sdk_core_1.ChainId.IMX_TESTNET;
        case 13371:
            return warp_sdk_core_1.ChainId.IMX_MAINNET;
        case 30732:
            return warp_sdk_core_1.ChainId.MOVEMENT_TESTNET;
        // new chain
        case 1300:
            return warp_sdk_core_1.ChainId.GLUE_TESTNET;
        default:
            throw new Error(`Unknown chain id: ${id}`);
    }
};
exports.ID_TO_CHAIN_ID = ID_TO_CHAIN_ID;
var ChainName;
(function (ChainName) {
    ChainName["MAINNET"] = "mainnet";
    ChainName["GOERLI"] = "goerli";
    ChainName["SEPOLIA"] = "sepolia";
    ChainName["OPTIMISM"] = "optimism-mainnet";
    ChainName["OPTIMISM_GOERLI"] = "optimism-goerli";
    ChainName["ARBITRUM_ONE"] = "arbitrum-mainnet";
    ChainName["ARBITRUM_GOERLI"] = "arbitrum-goerli";
    ChainName["POLYGON"] = "polygon-mainnet";
    ChainName["POLYGON_MUMBAI"] = "polygon-mumbai";
    ChainName["CELO"] = "celo-mainnet";
    ChainName["CELO_ALFAJORES"] = "celo-alfajores";
    ChainName["GNOSIS"] = "gnosis-mainnet";
    ChainName["MOONBEAM"] = "moonbeam-mainnet";
    ChainName["BNB"] = "bnb-mainnet";
    ChainName["AVALANCHE"] = "avalanche-mainnet";
    ChainName["BASE"] = "base-mainnet";
    ChainName["BASE_GOERLI"] = "base-goerli";
    ChainName["IMX_TESTNET"] = "imx-testnet";
    ChainName["IMX_MAINNET"] = "imx-mainnet";
    ChainName["MOVEMENT_TESTNET"] = "movement-testnet";
    // new chain
    ChainName["GLUE_TESTNET"] = "glue-testnet";
})(ChainName = exports.ChainName || (exports.ChainName = {}));
var NativeCurrencyName;
(function (NativeCurrencyName) {
    // Strings match input for CLI
    NativeCurrencyName["ETHER"] = "ETH";
    NativeCurrencyName["MATIC"] = "MATIC";
    NativeCurrencyName["CELO"] = "CELO";
    NativeCurrencyName["GNOSIS"] = "XDAI";
    NativeCurrencyName["MOONBEAM"] = "GLMR";
    NativeCurrencyName["BNB"] = "BNB";
    NativeCurrencyName["AVALANCHE"] = "AVAX";
})(NativeCurrencyName = exports.NativeCurrencyName || (exports.NativeCurrencyName = {}));
exports.NATIVE_NAMES_BY_ID = {
    [warp_sdk_core_1.ChainId.MAINNET]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.GOERLI]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.SEPOLIA]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.IMX_TESTNET]: [
        'IMX',
        'Immutable ZkEVM',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.IMX_MAINNET]: [
        'IMX',
        'Immutable ZkEVM',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.MOVEMENT_TESTNET]: [
        'MVM',
        'Movement Labs',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    // new chain
    [warp_sdk_core_1.ChainId.GLUE_TESTNET]: [
        'GLUE',
        'Glue',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.OPTIMISM]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.OPTIMISM_GOERLI]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.ARBITRUM_ONE]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.ARBITRUM_GOERLI]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.POLYGON]: ['MATIC', '0x0000000000000000000000000000000000001010'],
    [warp_sdk_core_1.ChainId.POLYGON_MUMBAI]: [
        'MATIC',
        '0x0000000000000000000000000000000000001010',
    ],
    [warp_sdk_core_1.ChainId.CELO]: ['CELO'],
    [warp_sdk_core_1.ChainId.CELO_ALFAJORES]: ['CELO'],
    [warp_sdk_core_1.ChainId.GNOSIS]: ['XDAI'],
    [warp_sdk_core_1.ChainId.MOONBEAM]: ['GLMR'],
    [warp_sdk_core_1.ChainId.BNB]: ['BNB', 'BNB', '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'],
    [warp_sdk_core_1.ChainId.AVALANCHE]: [
        'AVAX',
        'AVALANCHE',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
    [warp_sdk_core_1.ChainId.BASE]: [
        'ETH',
        'ETHER',
        '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    ],
};
exports.NATIVE_CURRENCY = {
    [warp_sdk_core_1.ChainId.MAINNET]: NativeCurrencyName.ETHER,
    [warp_sdk_core_1.ChainId.GOERLI]: NativeCurrencyName.ETHER,
    [warp_sdk_core_1.ChainId.SEPOLIA]: NativeCurrencyName.ETHER,
    [warp_sdk_core_1.ChainId.OPTIMISM]: NativeCurrencyName.ETHER,
    [warp_sdk_core_1.ChainId.OPTIMISM_GOERLI]: NativeCurrencyName.ETHER,
    [warp_sdk_core_1.ChainId.ARBITRUM_ONE]: NativeCurrencyName.ETHER,
    [warp_sdk_core_1.ChainId.ARBITRUM_GOERLI]: NativeCurrencyName.ETHER,
    [warp_sdk_core_1.ChainId.POLYGON]: NativeCurrencyName.MATIC,
    [warp_sdk_core_1.ChainId.POLYGON_MUMBAI]: NativeCurrencyName.MATIC,
    [warp_sdk_core_1.ChainId.CELO]: NativeCurrencyName.CELO,
    [warp_sdk_core_1.ChainId.CELO_ALFAJORES]: NativeCurrencyName.CELO,
    [warp_sdk_core_1.ChainId.GNOSIS]: NativeCurrencyName.GNOSIS,
    [warp_sdk_core_1.ChainId.MOONBEAM]: NativeCurrencyName.MOONBEAM,
    [warp_sdk_core_1.ChainId.BNB]: NativeCurrencyName.BNB,
    [warp_sdk_core_1.ChainId.AVALANCHE]: NativeCurrencyName.AVALANCHE,
    [warp_sdk_core_1.ChainId.BASE]: NativeCurrencyName.ETHER,
};
const ID_TO_NETWORK_NAME = (id) => {
    switch (id) {
        case 1:
            return ChainName.MAINNET;
        case 5:
            return ChainName.GOERLI;
        case 11155111:
            return ChainName.SEPOLIA;
        case 56:
            return ChainName.BNB;
        case 10:
            return ChainName.OPTIMISM;
        case 420:
            return ChainName.OPTIMISM_GOERLI;
        case 42161:
            return ChainName.ARBITRUM_ONE;
        case 421613:
            return ChainName.ARBITRUM_GOERLI;
        case 137:
            return ChainName.POLYGON;
        case 80001:
            return ChainName.POLYGON_MUMBAI;
        case 42220:
            return ChainName.CELO;
        case 44787:
            return ChainName.CELO_ALFAJORES;
        case 100:
            return ChainName.GNOSIS;
        case 1284:
            return ChainName.MOONBEAM;
        case 43114:
            return ChainName.AVALANCHE;
        case 8453:
            return ChainName.BASE;
        case 84531:
            return ChainName.BASE_GOERLI;
        case 13473:
            return ChainName.IMX_TESTNET;
        case 13371:
            return ChainName.IMX_MAINNET;
        case 30732:
            return ChainName.MOVEMENT_TESTNET;
        // new chain
        case 1300:
            return ChainName.GLUE_TESTNET;
        default:
            throw new Error(`Unknown chain id: ${id}`);
    }
};
exports.ID_TO_NETWORK_NAME = ID_TO_NETWORK_NAME;
exports.CHAIN_IDS_LIST = Object.values(warp_sdk_core_1.ChainId).map((c) => c.toString());
const ID_TO_PROVIDER = (id) => {
    switch (id) {
        case warp_sdk_core_1.ChainId.MAINNET:
            return 'https://rpc.ankr.com/eth';
        case warp_sdk_core_1.ChainId.GOERLI:
            return process.env.JSON_RPC_PROVIDER_GORLI;
        case warp_sdk_core_1.ChainId.SEPOLIA:
            return 'https://1rpc.io/sepolia';
        case warp_sdk_core_1.ChainId.OPTIMISM:
            return process.env.JSON_RPC_PROVIDER_OPTIMISM;
        case warp_sdk_core_1.ChainId.OPTIMISM_GOERLI:
            return process.env.JSON_RPC_PROVIDER_OPTIMISM_GOERLI;
        case warp_sdk_core_1.ChainId.ARBITRUM_ONE:
            return process.env.JSON_RPC_PROVIDER_ARBITRUM_ONE;
        case warp_sdk_core_1.ChainId.ARBITRUM_GOERLI:
            return process.env.JSON_RPC_PROVIDER_ARBITRUM_GOERLI;
        case warp_sdk_core_1.ChainId.POLYGON:
            return process.env.JSON_RPC_PROVIDER_POLYGON;
        case warp_sdk_core_1.ChainId.POLYGON_MUMBAI:
            return process.env.JSON_RPC_PROVIDER_POLYGON_MUMBAI;
        case warp_sdk_core_1.ChainId.CELO:
            return process.env.JSON_RPC_PROVIDER_CELO;
        case warp_sdk_core_1.ChainId.CELO_ALFAJORES:
            return process.env.JSON_RPC_PROVIDER_CELO_ALFAJORES;
        case warp_sdk_core_1.ChainId.BNB:
            return process.env.JSON_RPC_PROVIDER_BNB;
        case warp_sdk_core_1.ChainId.AVALANCHE:
            return process.env.JSON_RPC_PROVIDER_AVALANCHE;
        case warp_sdk_core_1.ChainId.BASE:
            return process.env.JSON_RPC_PROVIDER_BASE;
        case warp_sdk_core_1.ChainId.IMX_TESTNET:
            return 'https://wider-blue-wind.imx-testnet.quiknode.pro/3f38de94a68fde8c4c73946f2eb95cc6b831c6fa/';
        case warp_sdk_core_1.ChainId.IMX_MAINNET:
            return 'https://orbital-convincing-forest.imx-mainnet.quiknode.pro/c936618fee68fb1f858c0a4a5461d28dd7b880a2/';
        case warp_sdk_core_1.ChainId.MOVEMENT_TESTNET:
            return 'https://mevm.devnet.imola.movementlabs.xyz/';
        // new chain
        case warp_sdk_core_1.ChainId.GLUE_TESTNET:
            return 'https://testnet-ws-1.server-1.glue.net';
        default:
            throw new Error(`Chain id: ${id} not supported`);
    }
};
exports.ID_TO_PROVIDER = ID_TO_PROVIDER;
exports.WRAPPED_NATIVE_CURRENCY = {
    [warp_sdk_core_1.ChainId.MAINNET]: new warp_sdk_core_1.Token(1, '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.GOERLI]: new warp_sdk_core_1.Token(5, '0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.SEPOLIA]: new warp_sdk_core_1.Token(11155111, '0x7168Ee4968C867BD71e8E581f2f2b91F6B267ED0', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.IMX_TESTNET]: new warp_sdk_core_1.Token(13473, '0x7828DE82Bba71cc354aBd176Afea9F97afb59062', 18, 'WIMX', 'Wrapped IMX'),
    [warp_sdk_core_1.ChainId.IMX_MAINNET]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.IMX_MAINNET, '0x3A0C2Ba54D6CBd3121F01b96dFd20e99D1696C9D', 18, 'WIMX', 'Wrapped IMX'),
    [warp_sdk_core_1.ChainId.IMX_MAINNET]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.IMX_MAINNET, '0x3A0C2Ba54D6CBd3121F01b96dFd20e99D1696C9D', 18, 'WIMX', 'Wrapped IMX'),
    [warp_sdk_core_1.ChainId.MOVEMENT_TESTNET]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.MOVEMENT_TESTNET, '0xc02df8710Be33901D11A7E2D49B6c841e12B6f76', 18, 'WMVM', 'Wrapped MVM'),
    // new chain
    [warp_sdk_core_1.ChainId.GLUE_TESTNET]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.GLUE_TESTNET, '0x85256E9eA6F66d9B7677F1e9CE37594eCEf74424', 18, 'WGLUE', 'Wrapped GLUE'),
    [warp_sdk_core_1.ChainId.BNB]: new warp_sdk_core_1.Token(56, '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c', 18, 'WBNB', 'Wrapped BNB'),
    [warp_sdk_core_1.ChainId.OPTIMISM]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.OPTIMISM, '0x4200000000000000000000000000000000000006', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.OPTIMISM_GOERLI]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.OPTIMISM_GOERLI, '0x4200000000000000000000000000000000000006', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.ARBITRUM_ONE]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.ARBITRUM_ONE, '0x82aF49447D8a07e3bd95BD0d56f35241523fBab1', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.ARBITRUM_GOERLI]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.ARBITRUM_GOERLI, '0xe39Ab88f8A4777030A534146A9Ca3B52bd5D43A3', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.POLYGON]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.POLYGON, '0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270', 18, 'WMATIC', 'Wrapped MATIC'),
    [warp_sdk_core_1.ChainId.POLYGON_MUMBAI]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.POLYGON_MUMBAI, '0x9c3C9283D3e44854697Cd22D3Faa240Cfb032889', 18, 'WMATIC', 'Wrapped MATIC'),
    // The Celo native currency 'CELO' implements the erc-20 token standard
    [warp_sdk_core_1.ChainId.CELO]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.CELO, '0x471EcE3750Da237f93B8E339c536989b8978a438', 18, 'CELO', 'Celo native asset'),
    [warp_sdk_core_1.ChainId.CELO_ALFAJORES]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.CELO_ALFAJORES, '0xF194afDf50B03e69Bd7D057c1Aa9e10c9954E4C9', 18, 'CELO', 'Celo native asset'),
    [warp_sdk_core_1.ChainId.GNOSIS]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.GNOSIS, '0xe91d153e0b41518a2ce8dd3d7944fa863463a97d', 18, 'WXDAI', 'Wrapped XDAI on Gnosis'),
    [warp_sdk_core_1.ChainId.MOONBEAM]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.MOONBEAM, '0xAcc15dC74880C9944775448304B263D191c6077F', 18, 'WGLMR', 'Wrapped GLMR'),
    [warp_sdk_core_1.ChainId.AVALANCHE]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.AVALANCHE, '0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7', 18, 'WAVAX', 'Wrapped AVAX'),
    [warp_sdk_core_1.ChainId.BASE]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.BASE, '0x4200000000000000000000000000000000000006', 18, 'WETH', 'Wrapped Ether'),
    [warp_sdk_core_1.ChainId.BASE_GOERLI]: new warp_sdk_core_1.Token(warp_sdk_core_1.ChainId.BASE_GOERLI, '0x4200000000000000000000000000000000000006', 18, 'WETH', 'Wrapped Ether'),
};
function isMatic(chainId) {
    return chainId === warp_sdk_core_1.ChainId.POLYGON_MUMBAI || chainId === warp_sdk_core_1.ChainId.POLYGON;
}
class MaticNativeCurrency extends warp_sdk_core_1.NativeCurrency {
    equals(other) {
        return other.isNative && other.chainId === this.chainId;
    }
    get wrapped() {
        if (!isMatic(this.chainId))
            throw new Error('Not matic');
        const nativeCurrency = exports.WRAPPED_NATIVE_CURRENCY[this.chainId];
        if (nativeCurrency) {
            return nativeCurrency;
        }
        throw new Error(`Does not support this chain ${this.chainId}`);
    }
    constructor(chainId) {
        if (!isMatic(chainId))
            throw new Error('Not matic');
        super(chainId, 18, 'MATIC', 'Polygon Matic');
    }
}
function isCelo(chainId) {
    return chainId === warp_sdk_core_1.ChainId.CELO_ALFAJORES || chainId === warp_sdk_core_1.ChainId.CELO;
}
class CeloNativeCurrency extends warp_sdk_core_1.NativeCurrency {
    equals(other) {
        return other.isNative && other.chainId === this.chainId;
    }
    get wrapped() {
        if (!isCelo(this.chainId))
            throw new Error('Not celo');
        const nativeCurrency = exports.WRAPPED_NATIVE_CURRENCY[this.chainId];
        if (nativeCurrency) {
            return nativeCurrency;
        }
        throw new Error(`Does not support this chain ${this.chainId}`);
    }
    constructor(chainId) {
        if (!isCelo(chainId))
            throw new Error('Not celo');
        super(chainId, 18, 'CELO', 'Celo');
    }
}
function isGnosis(chainId) {
    return chainId === warp_sdk_core_1.ChainId.GNOSIS;
}
class GnosisNativeCurrency extends warp_sdk_core_1.NativeCurrency {
    equals(other) {
        return other.isNative && other.chainId === this.chainId;
    }
    get wrapped() {
        if (!isGnosis(this.chainId))
            throw new Error('Not gnosis');
        const nativeCurrency = exports.WRAPPED_NATIVE_CURRENCY[this.chainId];
        if (nativeCurrency) {
            return nativeCurrency;
        }
        throw new Error(`Does not support this chain ${this.chainId}`);
    }
    constructor(chainId) {
        if (!isGnosis(chainId))
            throw new Error('Not gnosis');
        super(chainId, 18, 'XDAI', 'xDai');
    }
}
function isBnb(chainId) {
    return chainId === warp_sdk_core_1.ChainId.BNB;
}
class BnbNativeCurrency extends warp_sdk_core_1.NativeCurrency {
    equals(other) {
        return other.isNative && other.chainId === this.chainId;
    }
    get wrapped() {
        if (!isBnb(this.chainId))
            throw new Error('Not bnb');
        const nativeCurrency = exports.WRAPPED_NATIVE_CURRENCY[this.chainId];
        if (nativeCurrency) {
            return nativeCurrency;
        }
        throw new Error(`Does not support this chain ${this.chainId}`);
    }
    constructor(chainId) {
        if (!isBnb(chainId))
            throw new Error('Not bnb');
        super(chainId, 18, 'BNB', 'BNB');
    }
}
function isMoonbeam(chainId) {
    return chainId === warp_sdk_core_1.ChainId.MOONBEAM;
}
class MoonbeamNativeCurrency extends warp_sdk_core_1.NativeCurrency {
    equals(other) {
        return other.isNative && other.chainId === this.chainId;
    }
    get wrapped() {
        if (!isMoonbeam(this.chainId))
            throw new Error('Not moonbeam');
        const nativeCurrency = exports.WRAPPED_NATIVE_CURRENCY[this.chainId];
        if (nativeCurrency) {
            return nativeCurrency;
        }
        throw new Error(`Does not support this chain ${this.chainId}`);
    }
    constructor(chainId) {
        if (!isMoonbeam(chainId))
            throw new Error('Not moonbeam');
        super(chainId, 18, 'GLMR', 'Glimmer');
    }
}
function isAvax(chainId) {
    return chainId === warp_sdk_core_1.ChainId.AVALANCHE;
}
class AvalancheNativeCurrency extends warp_sdk_core_1.NativeCurrency {
    equals(other) {
        return other.isNative && other.chainId === this.chainId;
    }
    get wrapped() {
        if (!isAvax(this.chainId))
            throw new Error('Not avalanche');
        const nativeCurrency = exports.WRAPPED_NATIVE_CURRENCY[this.chainId];
        if (nativeCurrency) {
            return nativeCurrency;
        }
        throw new Error(`Does not support this chain ${this.chainId}`);
    }
    constructor(chainId) {
        if (!isAvax(chainId))
            throw new Error('Not avalanche');
        super(chainId, 18, 'AVAX', 'Avalanche');
    }
}
class ExtendedEther extends warp_sdk_core_1.Ether {
    get wrapped() {
        if (this.chainId in exports.WRAPPED_NATIVE_CURRENCY) {
            return exports.WRAPPED_NATIVE_CURRENCY[this.chainId];
        }
        throw new Error('Unsupported chain ID');
    }
    static onChain(chainId) {
        var _a;
        return ((_a = this._cachedExtendedEther[chainId]) !== null && _a !== void 0 ? _a : (this._cachedExtendedEther[chainId] = new ExtendedEther(chainId)));
    }
}
exports.ExtendedEther = ExtendedEther;
ExtendedEther._cachedExtendedEther = {};
const cachedNativeCurrency = {};
function nativeOnChain(chainId) {
    if (cachedNativeCurrency[chainId] != undefined) {
        return cachedNativeCurrency[chainId];
    }
    if (isMatic(chainId)) {
        cachedNativeCurrency[chainId] = new MaticNativeCurrency(chainId);
    }
    else if (isCelo(chainId)) {
        cachedNativeCurrency[chainId] = new CeloNativeCurrency(chainId);
    }
    else if (isGnosis(chainId)) {
        cachedNativeCurrency[chainId] = new GnosisNativeCurrency(chainId);
    }
    else if (isMoonbeam(chainId)) {
        cachedNativeCurrency[chainId] = new MoonbeamNativeCurrency(chainId);
    }
    else if (isBnb(chainId)) {
        cachedNativeCurrency[chainId] = new BnbNativeCurrency(chainId);
    }
    else if (isAvax(chainId)) {
        cachedNativeCurrency[chainId] = new AvalancheNativeCurrency(chainId);
    }
    else {
        cachedNativeCurrency[chainId] = ExtendedEther.onChain(chainId);
    }
    return cachedNativeCurrency[chainId];
}
exports.nativeOnChain = nativeOnChain;
