// @warpgatex/warp-smart-order-router/build/main/providers/v2/static-subgraph-provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticV2SubgraphProvider = void 0;
const v2_sdk_1 = require("@uniswap/v2-sdk");
const warp_sdk_core_1 = require("@warpgatex/warp-sdk-core");
const lodash_1 = __importDefault(require("lodash"));
const chains_1 = require("../../util/chains");
const log_1 = require("../../util/log");
const token_provider_1 = require("../token-provider");
const token_provider_2 = require("./../token-provider");
const BASES_TO_CHECK_TRADES_AGAINST = {
    [warp_sdk_core_1.ChainId.MAINNET]: [
        chains_1.WRAPPED_NATIVE_CURRENCY[warp_sdk_core_1.ChainId.MAINNET],
        token_provider_1.DAI_MAINNET,
        token_provider_1.USDC_MAINNET,
        token_provider_1.USDT_MAINNET,
        token_provider_1.WBTC_MAINNET,
    ],
    [warp_sdk_core_1.ChainId.GOERLI]: [chains_1.WRAPPED_NATIVE_CURRENCY[warp_sdk_core_1.ChainId.GOERLI]],
    [warp_sdk_core_1.ChainId.SEPOLIA]: [chains_1.WRAPPED_NATIVE_CURRENCY[warp_sdk_core_1.ChainId.SEPOLIA]],
    [warp_sdk_core_1.ChainId.IMX_TESTNET]: [chains_1.WRAPPED_NATIVE_CURRENCY[warp_sdk_core_1.ChainId.IMX_TESTNET], token_provider_1.USDC_IMX_TESTNET],
    [warp_sdk_core_1.ChainId.IMX_MAINNET]: [chains_1.WRAPPED_NATIVE_CURRENCY[warp_sdk_core_1.ChainId.IMX_MAINNET], token_provider_1.USDC_IMX_MAINNET],
    [warp_sdk_core_1.ChainId.MOVEMENT_TESTNET]: [chains_1.WRAPPED_NATIVE_CURRENCY[warp_sdk_core_1.ChainId.MOVEMENT_TESTNET], token_provider_1.USDC_MOVEMENT_TESTNET],
    // new chain
    [warp_sdk_core_1.ChainId.GLUE_TESTNET]: [chains_1.WRAPPED_NATIVE_CURRENCY[warp_sdk_core_1.ChainId.GLUE_TESTNET], token_provider_2.USDC_GLUE_TESTNET],
    //v2 not deployed on [optimism, arbitrum, polygon, celo, gnosis, moonbeam, bnb, avalanche] and their testnets
    [warp_sdk_core_1.ChainId.OPTIMISM]: [],
    [warp_sdk_core_1.ChainId.ARBITRUM_ONE]: [],
    [warp_sdk_core_1.ChainId.ARBITRUM_GOERLI]: [],
    [warp_sdk_core_1.ChainId.OPTIMISM_GOERLI]: [],
    [warp_sdk_core_1.ChainId.POLYGON]: [],
    [warp_sdk_core_1.ChainId.POLYGON_MUMBAI]: [],
    [warp_sdk_core_1.ChainId.CELO]: [],
    [warp_sdk_core_1.ChainId.CELO_ALFAJORES]: [],
    [warp_sdk_core_1.ChainId.GNOSIS]: [],
    [warp_sdk_core_1.ChainId.MOONBEAM]: [],
    [warp_sdk_core_1.ChainId.BNB]: [],
    [warp_sdk_core_1.ChainId.AVALANCHE]: [],
    [warp_sdk_core_1.ChainId.BASE_GOERLI]: [],
    [warp_sdk_core_1.ChainId.BASE]: [],
};
/**
 * Provider that does not get data from an external source and instead returns
 * a hardcoded list of Subgraph pools.
 *
 * Since the pools are hardcoded, the liquidity/price values are dummys and should not
 * be depended on.
 *
 * Useful for instances where other data sources are unavailable. E.g. subgraph not available.
 *
 * @export
 * @class StaticV2SubgraphProvider
 */
class StaticV2SubgraphProvider {
    constructor(chainId) {
        this.chainId = chainId;
    }
    async getPools(tokenIn, tokenOut) {
        log_1.log.info('In static subgraph provider for V2');
        const bases = BASES_TO_CHECK_TRADES_AGAINST[this.chainId];
        const basePairs = lodash_1.default.flatMap(bases, (base) => bases.map((otherBase) => [base, otherBase]));
        if (tokenIn && tokenOut) {
            basePairs.push([tokenIn, tokenOut], ...bases.map((base) => [tokenIn, base]), ...bases.map((base) => [tokenOut, base]));
        }
        const pairs = (0, lodash_1.default)(basePairs)
            .filter((tokens) => Boolean(tokens[0] && tokens[1]))
            .filter(([tokenA, tokenB]) => tokenA.address !== tokenB.address && !tokenA.equals(tokenB))
            .value();
        const poolAddressSet = new Set();
        const subgraphPools = (0, lodash_1.default)(pairs)
            .map(([tokenA, tokenB]) => {
            const poolAddress = v2_sdk_1.Pair.getAddress(tokenA, tokenB);
            if (poolAddressSet.has(poolAddress)) {
                return undefined;
            }
            poolAddressSet.add(poolAddress);
            const [token0, token1] = tokenA.sortsBefore(tokenB)
                ? [tokenA, tokenB]
                : [tokenB, tokenA];
            return {
                id: poolAddress,
                liquidity: '100',
                token0: {
                    id: token0.address,
                },
                token1: {
                    id: token1.address,
                },
                supply: 100,
                reserve: 100,
                reserveUSD: 100,
            };
        })
            .compact()
            .value();
        return subgraphPools;
    }
}
exports.StaticV2SubgraphProvider = StaticV2SubgraphProvider;
